### PR TITLE
Update sub_windows order with respect to FLAG_ALWAYS_ON_TOP

### LINKED ISSUE #67106: Window's always_on_top flag doesn't seem to have any effect
### Godot version

4.0.beta2

### System information

Windows 11

### Issue description

According to the editor tooltip, if the always_on_top flag is set for a Window it should always stay on top of other windows (unless the transient flag is set). This does not seem to work. The window behaves the same way regardless of whether or not the always_on_top flag is set, even when the transient flag is disabled.

### Steps to reproduce

Note: I'm using PopupPanels here (and in the reproduction project) but the result is the same even with the base Window class.
Add a few PopupPanels to a scene. Set the popup_window, transient and borderless flags to false for all of them (so they behave like "normal" movable windows). Set always_on_top to true for one of them. Expected behavior is that that PopupPanel should always stay on top of the other ones, but it doesn't. It behaves the same way as the rest of them.

### Minimal reproduction project

[always_on_top_test.zip](https://github.com/godotengine/godot/files/9740520/always_on_top_test.zip)

### PR DESCRIPTION
**Pull request by ktxyz**
Fixes #67106.

### CODE REVIEW

**Comment by Sauermann** [comment]
Thanks for your PR.

In order to link this PR to the issue on the bug-tracker, you should include "fix #67106" in your opening post. It doesn't link to the issue, if you put it into the title.

You PR changes the display order correctly as far as I can tell, however it only tackles half of the problem, because input events also have to be adjusted. You would need to make sure, that input events get sent to the right window.

Currently I have the following problems:
- I can not move the ALWAYS_ON_TOP window around
- I move one of the lower windows around, when I click+drag within the top window

This video demonstrates it:
https://user-images.githubusercontent.com/6299227/198845590-e922a796-3f5b-418a-86d4-6a19e96e85c8.mp4

**Comment by ktxyz** [comment]
Sorry for all this commotion with this pull request. I found that order for both draw and input is set in `_sub_window_grab_focus`. This works in provided example and some others. Tested on Windows 10. Fixes #67106.

**Comment by kleonc** [comment]
> Sorry for all this commotion with this pull request. I found that order for both draw and input is set in `_sub_window_grab_focus`. This commit works in provided example and the original one. Tested on Windows 10. Fixes #67106.

No problem, just please update [the PR description](https://github.com/godotengine/godot/pull/68027#issue-1428357880) appropriately.

**Comment by ktxyz** [comment]
Most actions except for windows failed because of the strict `-Wno-unused`, I fixed it and will reopen the pull request when fork actions finish with success.

**Comment by Sauermann** [comment]
I like the idea to sort the order on insertion.

You should also have a look at `Viewport::_sub_window_register`
https://github.com/godotengine/godot/blob/aaa931659e734953f649e229ca587b5b64703a02/scene/main/viewport.cpp#L185
This is another location where subwindows are pushed into the list and that would need to be adjusted.

**Comment by ktxyz** [comment]
Yeah, the `Viewport::_sub_window_update_order` method seems to be a good place to put the reordering into since its used in this exact context.
https://github.com/ktxyz/godot/blob/91bebf444bbc0a972807246ca9b96cc11751265b/scene/main/viewport.cpp#L164-L184

The `Viewport::_sub_window_register` calls the `Viewport::_sub_window_grab_focus` at the end 
https://github.com/godotengine/godot/blob/aaa931659e734953f649e229ca587b5b64703a02/scene/main/viewport.cpp#L185-L187
but now that I thought about it does make sense to not assume that every registered window will grab focus also in the future.

File: scene/main/viewport.cpp
Comment (Sauermann) [inline scene/main/viewport.cpp:165]: You can safely return here, if the size of the list is zero or one.

File: scene/main/viewport.cpp
Comment (Sauermann) [inline scene/main/viewport.cpp:178]: If `index == gui.sub_windows.size() - 1`, you remove the element and place it back to the same location.

**Comment by Sauermann** [comment]
The functionality looks good to me, it seems to work as it should.

File: scene/main/viewport.cpp
Comment (ktxyz) [reply scene/main/viewport.cpp]: Updated the commit

File: scene/main/viewport.cpp
Comment (ktxyz) [reply scene/main/viewport.cpp]: Added if statement

File: scene/main/viewport.cpp
Comment (kleonc) [inline scene/main/viewport.cpp:166]:
```suggestion
	if (gui.sub_windows.size() < 2) {
		return;
	}
```

File: scene/main/viewport.cpp
Comment (kleonc) [inline scene/main/viewport.cpp:169]:
```suggestion
	if (!gui.sub_windows[index].window->get_flag(Window::FLAG_ALWAYS_ON_TOP)) {
```

File: scene/main/viewport.cpp
Comment (kleonc) [inline scene/main/viewport.cpp:175]: Not obligatory but I think it would be simpler:
```suggestion
		while (index > 0 && gui.sub_windows[index - 1].window->get_flag(Window::FLAG_ALWAYS_ON_TOP)) {
			--index;
		}
```

**Review by kleonc — COMMENTED** [review]
Added some code style related suggestions.

**Comment by ktxyz** [comment]
Fixed code style and swapped for FOR while.

File: scene/main/viewport.cpp
Comment (Sauermann) [inline scene/main/viewport.cpp:171]: `sw` is only used within the if-block below. You could move the declaration down there.

Comment (ktxyz) [reply scene/main/viewport.cpp]: alright, also moved `index` initialization inside the if

File: scene/main/viewport.cpp
Comment (Sauermann) [inline scene/main/viewport.cpp:177]: The array index should be `gui.sub_windows.size() - 1`

Comment (ktxyz) [reply scene/main/viewport.cpp]: Sorry about that, still learning how git works exactly(I did git add before changing this line, then changed it and commited, my bad). Fixed.

Comment (Sauermann) [reply scene/main/viewport.cpp]: No worries. It takes a bit to get used to git.

**Comment by ktxyz** [comment]
So what now? Will it be looked at later? I see it's a low priority bug. Sorry for asking but my it's my first pull request.

**Comment by Sauermann** [comment]
Yes, patience is beneficial. The list of open PRs is much longer than the list of people who can approve PRs.

**Comment by kleonc** [comment]
> So what now? Will it be looked at later?

@ktxyz This PR as well as many other window/input/transform/etc. related PRs from @Sauermann need to be reviewed/approved by @reduz as it's "his area". I can tell he's aware of these and has planned to review/tackle them at some point. So yes, just a matter of time (can't estimate though, maybe in a week, maybe in a month, maybe longer).

@Sauermann BTW I think I haven't seen you approving a PR (I might be wrong). So just in case: don't hesitate to approve PRs if you have no doubts about their correctness. :slightly_smiling_face:

**Comment by Sauermann** [comment]
@kleonc No, I don't have the power to approve pull requests. But after reviewing this bugfix, I would implement it exactly in the same way. (ok, maybe change `--index;` to `index--;`, but that is my personal preference and changes nothing in the functionality)

**Comment by kleonc** [comment]
> No, I don't have the power to approve pull requests.

Oh, [you do have such power](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/approving-a-pull-request-with-required-reviews). The difference is your (non-member) approval won't e.g. unblock merging (and will be shown in gray instead of in green). But it's still a valid feedback for the maintainers (just like any other form of reviewing), especially if they do think you're credible (based on your past reviews/PRs/etc.).

(end of off topic :wink:)

**Review by Sauermann — APPROVED** [review]

**Review by bruvzg — APPROVED** [review]

**Comment by akien-mga** [comment]
Thanks! And congrats for your first merged Godot contribution :tada:

**Pull request merged**
